### PR TITLE
WIP: Capture OS distro info in Javacore

### PR DESCRIPTION
**Pull request by subbachari**
In the processor/GPINFO section of javacore, capture the linux distribution information which will be helpful during debugging.
The Sample output will be like this
NULL           ------------------------------------------------------------------------
0SECTION       GPINFO subcomponent dump routine
NULL           ================================
2XHOSLEVEL     OS Level         : Linux 4.4.84-boot2docker
2XHOSDISTRO     OS Distribution         : "Ubuntu 17.04"
2XHCPUS        Processors -
3XHCPUARCH       Architecture   : amd64
3XHNUMCPUS       How Many       : 1
3XHNUMASUP       NUMA is either not supported or has been disabled by user
NULL
1XHERROR2      Register dump section only produced for SIGSEGV, SIGILL or SIGFPE.
NULL
NULL           ------------------------------------------------------------------------
Signed-off-by: subrahmanya chari p <subbachari@in.ibm.com>

### CODE REVIEW

**Comment by mgaudet** [comment]
It appears something may have gone wrong with a rebase you did, as you appear to have two different commits with identical commit messages; if they are part of the same pieces of work it may be worth squashing them into one commit.

Regarding your PR Description: You can make your text show up preformatted by either fencing it with three backticks, or spacing with four spaces: 

    ```
    fenced code shows up pre
    ```

```
    four spaces also 
    leads to pre
```

For `inline pre`, wrap the section with a single backtick: 

    Lorem ipsum `<html>` dolor sit amet

See more in the GitHub [Guide to Mastering Markdown](https://guides.github.com/features/mastering-markdown/)

File: runtime/rasdump/javadump.cpp [hunk 1]
```diff
@@ -89,6 +111,42 @@ static jvmtiIterationControl regionIteratorCallback (J9JavaVM* vm, J9MM_IterateR
...
 static UDATA getAllocatedVMThreadCount (J9JavaVM *vm);
 
+
+/**
+ * This function is used to get the Redhat OS distribution details
+ * @param [out] OS Distribution Name
+ *
+ * @return returns 0 on success, any other -1 code on failure
+ */
+
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: Please omit the blank line between the description and the forward declaration.
Also, please leave a blank line before the next desciption.
This should reduce confusion about how comments are paired with those declarations.

File: runtime/rasdump/javadump.cpp [hunk 2]
```diff
@@ -89,6 +111,42 @@ static jvmtiIterationControl regionIteratorCallback (J9JavaVM* vm, J9MM_IterateR
...
+static int getUbuntuDistributionInfo(char *strDistributionVersion);
+/**
+ * This function is used to get theOS distribution details
+ * @param [out] OS Distribution Name
+ *
+ * @return returns 0 on success, any other -1 code on failure
+ */
+
+static int getOsDistributionName(char *strOSdirstibution);
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: Please fix the typo: strOSdirstibution should be strOSdistribution.

File: runtime/rasdump/javadump.cpp [hunk 3]
```diff
@@ -849,6 +907,9 @@ JavaCoreDumpWriter::writeProcessorSection(void)
 		"NULL           ================================\n"
 	);
 
+	char strOsDirstibution[OPENJ9_VERSION_INFO_SIZE] = "";
+    getOsDistributionName(strOsDirstibution);
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: Please indent with tabs as is done in the nearby code.

File: runtime/rasdump/javadump.cpp [hunk 4]
```diff
@@ -5796,3 +5863,122 @@ getAllocatedVMThreadCount(J9JavaVM *vm)
...
+ * @return returns 0 on success, any other -1 code on failure
+ */
+static int 
+getRedhatDistributionInfo(char *strDistributionVersion)
+{
+	int nSuccess = -1;
+	char strLine [OPENJ9_VERSION_INFO_SIZE] = "";
+	FILE *fpVersion = NULL;
+	/*read redhat Os distribution info file */
+	fpVersion = fopen(OPENJ9_REDHAT_OS_DISTRIBUTION_INFO_FILE, "r");
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: Please use OMR support API to make the code more portable:
```
fopen  -> omrfile_open
fgets  -> omrfile_read
fclose -> omrfile_close
```
The file must be closed regardless of whether the read was successful.

File: runtime/rasdump/javadump.cpp [hunk 5]
```diff
@@ -5796,3 +5863,122 @@ getAllocatedVMThreadCount(J9JavaVM *vm)
...
+static int 
+getUbuntuDistributionInfo(char *strDistributionVersion)
+{
+	int nSuccess = -1;
+	char strLine[OPENJ9_VERSION_INFO_SIZE]="";
+	FILE *fpVersion=NULL;
+	fpVersion = fopen(OPENJ9_UBUNTU_OS_DISTRIBUTION_INFO_FILE, "r");
+	if (fpVersion!=NULL) {
+		while (NULL != fgets(strLine, OPENJ9_VERSION_INFO_SIZE, fpVersion)) {
+			if (NULL != strstr(strLine, OPENJ9_PRETTY_NAME) ) {
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: When `strstr` returns a non-NULL value, it points at the desired string. There's no need to use `strtok` to find what follows, simply capture the return value and add the length of the prefix to find the value.

Please be consistent with formatting (the functions above are better). For example, there should be spaces either side of operators like `=` and `!=`; NULL should be on the left of comparisons; a space should appear after each comma.

File: runtime/rasdump/javadump.cpp [hunk 6]
```diff
@@ -5796,3 +5863,122 @@ getAllocatedVMThreadCount(J9JavaVM *vm)
...
+	return nSuccess;
+}
+/**
+ * This function is used to get theOS distribution details
+ * @param [out] OS Distribution Name
+ *
+ * @return returns 0 on success, any other -1 code on failure
+ */
+static int 
+getOsDistributionName(char *strOSdirstibution)
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: Please fix the typo: strOSdirstibution should be strOSdistribution.

File: runtime/rasdump/javadump.cpp [hunk 7]
```diff
@@ -5796,3 +5863,122 @@ getAllocatedVMThreadCount(J9JavaVM *vm)
...
+{
+	int nSuccess = -1;
+	int nDistributionType = 0;
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp]: Please use a name instead of `0`, e.g. `OPENJ9_UNKNOWN_DISTRIBUTION`.

**Comment by keithc-ca** [comment]
The proposed changes seem only applicable to Linux, and within that only handles three distributions.
The Linux-specific code should be marked as such (`if defined(LINUX)`)?
Can other distributions be recognized generically?

OpenJ9 is intended for use on many other platforms (e.g. Windows, z/OS): is this only relevant on Linux?

**Comment by joransiu** [comment]
> OpenJ9 is intended for use on many other platforms (e.g. Windows, z/OS): is this only relevant on Linux?

For z/OS, we already show the OS level: i.e. "OS Level" like "z/OS 02.02.00".  However, we do want the underlying H/W levels and capabilities in the javacore.  See https://github.com/eclipse/openj9/issues/94

**Comment by ashu-mehra** [comment]
I am wondering if these APIs are better off in port library?

**Comment by DanHeidinga** [comment]
> I am wondering if these APIs are better off in port library?

+1.  The get the OS distribution APIs would be a good addition to the OMR port library, likely in the `sysinfo` area.

**Comment by subbachari** [comment]
Thanks for your suggestions , I am working on them, will update once I made the progress

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by keithc-ca** [comment]
This hasn't progressed in 3.5 years.

**Pull request closed without merging**